sdk/metric/meter_test.go

```diff
@@ -64,3 +72,452 @@ func TestMeterRegistry(t *testing.T) {
+
+			inst := tt.genInstrument(m)
+			err := m.RegisterCallback([]instrument.Asynchronous{inst}, func(ctx context.Context) {
+				inst.Observe(ctx, 5)
+			})
+			assert.NoError(t, err)
+
+			inst.Observe(context.Background(), 7)
+
+			rm, err := rdr.Collect(context.Background())
+			assert.NoError(t, err)
+
+			require.Len(t, rm.ScopeMetrics, 1)
+			sm := rm.ScopeMetrics[0]
+			require.Len(t, sm.Metrics, 1)
+			got := sm.Metrics[0]
+			metricdatatest.AssertEqual(t, tt.want, got, metricdatatest.IgnoreTimestamp())
+		})
+	}
+}
+
+func testFloatAsyncCallback(t *testing.T) {
+	type Observer interface {
+		instrument.Asynchronous
+		Observe(context.Context, float64, ...attribute.KeyValue)
+	}
+	testCases := []struct {
+		name          string
+		genInstrument func(metric.Meter) Observer
+		want          metricdata.Metrics
+	}{
+		{
+			name: "Counter",
+			genInstrument: func(m metric.Meter) Observer {
+				ctr, _ := m.AsyncFloat64().Counter("counter")
+				return ctr
+			},
+			want: metricdata.Metrics{
+				Name: "counter",
+				Data: metricdata.Sum[float64]{
+					Temporality: metricdata.CumulativeTemporality,
+					IsMonotonic: true,
+					DataPoints: []metricdata.DataPoint[float64]{
+						{Value: 5},
+					},
+				},
+			},
+		},
+		{
+			name: "UpDownCounter",
+			genInstrument: func(m metric.Meter) Observer {
+				ctr, _ := m.AsyncFloat64().UpDownCounter("UpDownCounter")
+				return ctr
+			},
+			want: metricdata.Metrics{
+				Name: "UpDownCounter",
+				Data: metricdata.Sum[float64]{
+					Temporality: metricdata.CumulativeTemporality,
+					IsMonotonic: false,
+					DataPoints: []metricdata.DataPoint[float64]{
+						{Value: 5},
+					},
+				},
+			},
+		},
+		{
+			name: "Gauge",
+			genInstrument: func(m metric.Meter) Observer {
+				ctr, _ := m.AsyncFloat64().Gauge("Gauge")
+				return ctr
+			},
+			want: metricdata.Metrics{
+				Name: "Gauge",
+				Data: metricdata.Gauge[float64]{
+					DataPoints: []metricdata.DataPoint[float64]{
+						{Value: 5},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			rdr := NewManualReader()
+			m := NewMeterProvider(WithReader(rdr)).Meter("testInstruments")
+
+			inst := tt.genInstrument(m)
+			err := m.RegisterCallback([]instrument.Asynchronous{inst}, func(ctx context.Context) {
+				inst.Observe(ctx, 5)
+			})
+			assert.NoError(t, err)
+
+			inst.Observe(context.Background(), 7)
+
+			rm, err := rdr.Collect(context.Background())
+			assert.NoError(t, err)
+
+			require.Len(t, rm.ScopeMetrics, 1)
+			sm := rm.ScopeMetrics[0]
+			require.Len(t, sm.Metrics, 1)
+			got := sm.Metrics[0]
+			metricdatatest.AssertEqual(t, tt.want, got, metricdatatest.IgnoreTimestamp())
+		})
+	}
+}
+
+// Async Instruments should not be usable outside of callback.
+func TestAsyncInstrumentsWithinCallback(t *testing.T) {
+	t.Run("Int64", testIntAsyncCallback)
+	t.Run("Float64", testFloatAsyncCallback)
```

Comment:
`testIntAsyncCallback` and `testFloatAsyncCallback` do more than check that instruments are not usable outside of callbacks, the additional testing they do duplicates what is done in `TestMeterCreatesInstruments`. These tests need to be made to not overlap

Reply:
I agree they are very similar, and that's because of what was chosen as a the Unit Under Test (UUT).  In both cases, we need the logic of the registry within the meter to be able to see outputs that are produced.

The latter test `TestMeterCreatesInstruments` is intended to test (eventually) all instruments generate the correct aggregation in the output.

The former tests that an observe call made outside of the `produce()`, triggered by `Collect()`, are not recorded in the output.  Because of that boundary, I can't test this logic without also testing that it is the correct aggregation.  

If you want to reduce the duplication of tests I could put the spurious `Observe()` into the `TestMeterCreatesInstruments`.  That would overload that test to be both about the correctness of the aggregation and the non-callback discards.